### PR TITLE
Fixes #223: test case added for deploy script

### LINKED ISSUE #223: send heroku deployment link in mail
### This is a
- [ ] Bug report
- [x] Feature request

### Current behavior

now we sending only GitHub deploy link
### Expected behavior

send Heroku deployment link as well in the mail

### Steps to reproduce
1.
2.
3.

### PR DESCRIPTION
**Pull request by sch00lb0y**
## Description

I have added test case for deploy script
## Types of changes

- [ ] Bug fix 
- [x] New feature

## Checklist:

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] There is a corresponding issue for this pull request.
- [x] Mentioned the Issue number in the pull request commit message `Fixes #<number> commit message`
- [x] There is only one commit per issue.

### CODE REVIEW

File: backend/deploy.js [hunk 1]
```diff
@@ -1,11 +1,12 @@
 var crypter = require("../util/crypter.js");
 var spawn = require('child_process').spawn;
 var output = require('../util/output');
+var socketHandler = require('../util/socketHandler.js');
 
-exports.deployPages = function (socket, data) {
+exports.deployPages = function (socket, data, callback) {
```
Comment (pri22296) [inline backend/deploy.js]: Why are these changes made in a PR about adding test cases?

Comment (sch00lb0y) [inline backend/deploy.js]: @pri22296 I agree with your query but then those are not pure function how will write test case if the function doesn't return anything. That's why I added a callback. if you have any workaround to test this kind of function please suggest me.

Comment (pri22296) [inline backend/deploy.js]: adding callbacks to code just for the sake of testing seems like a bad idea. If the function doesn't return anything, try to test the side effects. Like if you want to test that a method is called, maybe mock that method using any mock testing framework.

Comment (sch00lb0y) [inline backend/deploy.js]: I have rewritten the test with fake socket object. please check

**Comment by sch00lb0y** [comment]
@imujjwal96 please check this.Once you reviewers okay with PR. I'll add the necessary environmental variable to Travis

File: tests/test.js [hunk 2]
```diff
@@ -14,28 +15,67 @@ describe('crypter test', () => {
...
+      if (err) {
+        generatorExitCode = err.exitCode;
+        done()
+      } else {
+        uniqueId = generatedData.uniqueId;
+        generatorExitCode = generatedData.exitCode;
+        files = fs.readdirSync(`temp/${email}`);
+        deploy.deployPages(fakeSocket(done), {
+          gitURL: "https://github.com/sch00lb0y/yaydoc.git",
+          encryptedToken: crypter.encrypt(process.env.GITHUB_OAUTH_TOKEN),
```
Comment (imujjwal96) [inline tests/test.js]: Is this the actual Github token?

File: tests/test.js [hunk 3]
```diff
@@ -14,28 +15,67 @@ describe('crypter test', () => {
...
+          email: email,
+          uniqueId: uniqueId,
+          username: "sch00lb0y"
+        })
+        deploy.deployHeroku(fakeSocket(done), {
+          email: email,
+          herokuAppName: "yaydoctest",
+          herokuAPIKey: crypter.encrypt(process.env.HEROKU_API_KEY),
```
Comment (imujjwal96) [inline tests/test.js]: the original API KEY?

File: tests/test.js (hunk 2 above)
Comment (sch00lb0y) [inline tests/test.js]: yeah, that's why I'm testing in my local repository. In [open-event-android](https://github.com/fossasia/open-event-android/blob/development/upload-apk.sh#L15) also they been using GitHub token to push apk

File: tests/test.js (hunk 3 above)
Comment (sch00lb0y) [inline tests/test.js]: yes

File: tests/test.js (hunk 2 above)
Comment (imujjwal96) [inline tests/test.js]: can you not mock this value here, as this is just a test case?

File: tests/test.js [hunk 4]
```diff
@@ -14,28 +15,67 @@ describe('crypter test', () => {
...
-      done()
+    generator.executeScript({}, data, function(err, generatedData) {
+      if (err) {
+        generatorExitCode = err.exitCode;
+        done()
+      } else {
+        uniqueId = generatedData.uniqueId;
+        generatorExitCode = generatedData.exitCode;
+        files = fs.readdirSync(`temp/${email}`);
+        deploy.deployPages(fakeSocket(done), {
```
Comment (imujjwal96) [inline tests/test.js]: is this performing a deployment or are you mocking this function?

Comment (sch00lb0y) [inline tests/test.js]: I'm performing the deployment. I mocked socket object to get exitCode. then I'm asserting using that

File: tests/test.js (hunk 2 above)
Comment (sch00lb0y) [inline tests/test.js]: Actually, I can mock the function but we can't check the deployment. yeah, it is just a test case right. I'll try what I can do. Thanks for your suggestion. :smile: :+1: .

File: tests/test.js (hunk 4 above)
Comment (imujjwal96) [inline tests/test.js]: Please try not to perform a deployment in running a test case.

Comment (sch00lb0y) [inline tests/test.js]: @imujjwal96 please check the current implementation

**Comment by imujjwal96** [comment]
@sch00lb0y Why's the travis build failing?

File: tests/test.js [hunk 5]
```diff
@@ -1,50 +1,120 @@
...
+  }
+  if (process === undefined) {
+    throw Error("process object is not passed in the lineOutput");
+  }
+  if (event === undefined) {
+    throw Error("event object is not passed in the lineOutput");
+  }
+  if (increment === undefined) {
+    throw Error("increment object is not passed in the lineOutput");
+  }
```
Comment (imujjwal96) [inline tests/test.js]: - increment has to be an integer value
- increment cannot be <= 0 
- increment cannot be > 100

File: tests/test.js [hunk 6]
```diff
@@ -1,50 +1,120 @@
 require('dotenv').config({ path: './.env'});
```
Comment (imujjwal96) [inline tests/test.js]: distribute to tests to various files
like`tests/util-test.js`, `tests/backend-test.js` etc. I am not sure about the naming convention though. Please refer any relevant resource for that

Comment (sch00lb0y) [inline tests/test.js]: made changes please check

File: tests/backend-test.js [hunk 7]
```diff
@@ -0,0 +1,121 @@
+require('dotenv').config({ path: './.env'});
+const sinon = require("sinon");
+const assert = require('assert');
+const crypter = require('../util/crypter');
+sinon.stub(require("child_process"), "spawn").callsFake(function (fileName, args) {
+  if (fileName !== "./ghpages_deploy.sh" && fileName !== "./heroku_deploy.sh" && fileName !== "./generate.sh") {
+    throw new Error(`invalid ${fileName} invoked`);
+  }
+  if (fileName === "./ghpages_deploy.sh") {
+    let ghArgs = ["-e", "-i", "-w", "-n", "-o", "-r"];
```
Comment (pri22296) [inline tests/backend-test.js]: Remove the `-w` option as it has been removed

File: tests/backend-test.js [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    ghArgs.forEach(function (x) {
+      if (args.indexOf(x) < 0) {
+        throw new Error(`${x} argument is not passed`);
+      }
+    })
+  }
+  if (fileName === "./heroku_deploy.sh") {
+    let herokuArgs = ["-e", "-u", "-w", "-h", "-n"];
```
Comment (pri22296) [inline tests/backend-test.js]: here too

File: tests/backend-test.js [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    herokuArgs.forEach(function (x) {
+      if (args.indexOf(x) < 0) {
+        throw new Error(`${x} argument is not passed`);
+      }
+    });
+  }
+  if (fileName === "./generate.sh") {
+    let generateArgs = ["-g", "-t", "-m", "-d", "-u", "-w", "-s", "-p"];
```
Comment (pri22296) [inline tests/backend-test.js]: here too

Comment (sch00lb0y) [inline tests/backend-test.js]: made changes

**Comment by sch00lb0y** [comment]
@imujjwal96 please check this once again

**Pull request merged**